### PR TITLE
feat: rpc endpoint 'eth_newFilter'

### LINKED ISSUE #425: Implement `eth_newFilter` rpc endpoints
See: https://github.com/ethereum/execution-apis/blob/71260dde2d7fc5a12b9d171692a19a5e599820cf/src/eth/filter.yaml#L1C9-L22

### PR DESCRIPTION
**Pull request by fkrause98**
**Description**

- Implementation of the eth_newFilter RPC endpoint.
- This will 'install' a filter, which can later be polled with 'eth_getFilterChanges', yet to be implemented.
- Create the filter type to re use it both on rpc and storage crates.
- Added smoke tests since hive does not provide tests for eth_newFilter.

Closes #425

### CODE REVIEW

File: crates/storage/engines/libmdbx.rs [hunk 1]
```diff
@@ -461,6 +460,15 @@ table!(
     ( Payloads ) u64 => BlockRLP
 );
 
+// Current filters
+table!(
```
Comment (jrchatruc) [inline crates/storage/engines/libmdbx.rs:464]: Why use an `libmdbx` table for this instead of simply an in-memory hash map? I don't think we care much about persisting filters

File: crates/rpc/eth/filter.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        Ok(FilterRequest {
+            request_data: filter,
+        })
+    }
+    fn handle(
+        &self,
+        storage: ethereum_rust_storage::Store,
+    ) -> Result<serde_json::Value, crate::utils::RpcErr> {
+        let filter = self.request_data.request_to_filter(&storage)?;
+        let id: u64 = random();
```
Comment (mpaulucci) [inline crates/rpc/eth/filter.rs:27]: Could you investigate geth and see if they use a random key or a deterministic one? 

As an example, see how they generate the payload id: https://github.com/lambdaclass/ethereum_rust/pull/540/files#diff-4446e6639c57085d9923271c8e951db627558bd79daefc708351b4ccd2281e53R29

File: crates/core/types/filters.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+#[derive(Deserialize, Debug, Clone)]
+#[serde(untagged)]
+/// Topic filter used to filter Logs.
+pub enum TopicFilter {
+    Topic(H256),
+    Topics(Vec<H256>),
+}
+
+#[derive(Debug, Clone)]
+pub struct LogsFilter {
```
Comment (mpaulucci) [inline crates/core/types/filters.rs:27]: is this structure equivalent to the one you're using for getLogs?

File: crates/rpc/eth/filter.rs (hunk 2 above)
Comment (fkrause98) [reply crates/rpc/eth/filter.rs:27]: Sure, I'll doble check this.

File: crates/core/types/filters.rs (hunk 3 above)
Comment (fkrause98) [reply crates/core/types/filters.rs:27]: It's similar, I had to move it here to avoiding having circular dependencies, also I removed the options to declutter the code which will use this structure.

Comment (fkrause98) [reply crates/core/types/filters.rs:27]: I think I'll remove it, since now we'll be using a hashmap instead of using the db, and I won't need it for the storage crate.

File: crates/rpc/eth/filter.rs (hunk 2 above)
Comment (fkrause98) [reply crates/rpc/eth/filter.rs:45]: Just for the record, the ids are random:
- [filter subscription definition](https://github.com/ethereum/go-ethereum/blob/82b0dec7135b281c1b03064d50959dc992c2f94f/eth/filters/filter_system.go#L240-L245).
- [RPC id type definition](https://github.com/ethereum/go-ethereum/blob/15fb0dcc6732f3b04f8d3b13e309afd723364ba5/rpc/subscription.go#L50-L58)

File: crates/core/types/filters.rs (hunk 3 above)
Comment (fkrause98) [reply crates/core/types/filters.rs]: Removed!

File: crates/storage/engines/libmdbx.rs (hunk 1 above)
Comment (fkrause98) [reply crates/storage/engines/libmdbx.rs]: Done!

File: crates/storage/store/engines/libmdbx.rs [hunk 4]
```diff
@@ -558,7 +552,6 @@ pub fn init_db(path: Option<impl AsRef<Path>>) -> Database {
         table_info!(StateTrieNodes),
         table_info!(StorageTriesNodes),
         table_info!(CanonicalBlockHashes),
-        table_info!(Payloads),
```
Comment (mpaulucci) [inline crates/storage/store/engines/libmdbx.rs:561]: why are you removing this?

File: crates/storage/store/storage.rs [hunk 5]
```diff
@@ -5,13 +5,16 @@ use self::engines::libmdbx::Store as LibmdbxStore;
 use self::error::StoreError;
```
Comment (mpaulucci) [inline crates/storage/store/storage.rs:5]: why these changes here?

File: crates/networking/rpc/utils.rs [hunk 6]
```diff
@@ -209,3 +209,21 @@ fn get_message_from_revert_data(_data: &str) -> String {
     // See https://github.com/ethereum/go-ethereum/blob/8fd43c80132434dca896d8ae5004ae2aac1450d3/accounts/abi/abi.go#L275
     "".to_owned()
 }
+
+#[cfg(test)]
```
Comment (mpaulucci) [inline crates/networking/rpc/utils.rs:213]: where does this come from?

File: crates/storage/store/engines/libmdbx.rs (hunk 4 above)
Comment (fkrause98) [reply crates/storage/store/engines/libmdbx.rs]: My bad, [restored](https://github.com/lambdaclass/ethereum_rust/pull/567/commits/df1bf9b265c15d810cf608539fa6acd18bd50e37)

File: crates/networking/rpc/rpc.rs [hunk 7]
```diff
@@ -56,6 +62,7 @@ pub struct RpcApiContext {
     storage: Store,
     jwt_secret: Bytes,
     local_p2p_node: Node,
+    active_filters: ActiveFilters,
```
Comment (mpaulucci) [inline crates/networking/rpc/rpc.rs:65]: this is good for now, but maybe eventually we need to have a different approach than having a random list of things that some endpoints might need. but def out of scope of this PR.

**Review by mpaulucci — APPROVED** [review]
LGTM

File: crates/storage/store/storage.rs (hunk 5 above)
Comment (fkrause98) [reply crates/storage/store/storage.rs]: My bad, seems like I did the merge with main wrongly, storage + engine shouldn't have any changes now.

File: crates/networking/rpc/rpc.rs (hunk 7 above)
Comment (fkrause98) [reply crates/networking/rpc/rpc.rs:65]: I agree, we could leverage Axum's tools for this, but maybe on another issue/PR.
